### PR TITLE
Implement GPU signed scalar div, add scalar div to the hl api

### PR DESCRIPTION
**Pull request by agnesLeroy**
closes: _please link all relevant issues_

### Check-list:

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

File: tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs [hunk 1]
```diff
@@ -403,4 +422,262 @@ impl CudaServerKey {
...
+            );
+        }
+
+        let quotient;
+        if absolute_divisor == (Scalar::Unsigned::ONE << chosen_multiplier.l as usize) {
+            // Issue q = SRA(n + SRL(SRA(n, l − 1), N − l), l);
+            let l = chosen_multiplier.l;
+
+            // SRA(n, l − 1)
+            let mut tmp = self.unchecked_scalar_right_shift(numerator, l - 1, streams);
```
Comment (pdroalves) [inline tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs:478]: We should be using `unchecked_scalar_right_shift_async`.

File: tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs [hunk 2]
```diff
@@ -32,6 +36,21 @@ impl CudaServerKey {
...
+        &self,
+        lhs: &CudaSignedRadixCiphertext,
+        rhs: Scalar,
+        streams: &CudaStreams,
+    ) -> CudaSignedRadixCiphertext
+    where
+        Scalar: ScalarMultiplier + DecomposableInto<u8> + CastInto<u64>,
+    {
+        let num_blocks = lhs.as_ref().d_blocks.lwe_ciphertext_count().0;
+        let mut result = self.extend_radix_with_sign_msb(lhs, num_blocks, streams);
```
Comment (pdroalves) [inline tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs:49]: Not directly related to this PR but still in the scope,`extend_radix_with_sign_msb` is not following the async/sync standards. There should exist `extend_radix_with_sign_msb_assign_async` and `extend_radix_with_sign_msb_async` so we use the latter here. The same for `signed_scalar_mul_high`, we should have an async variant and use it in `unchecked_signed_scalar_div`.

File: tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs [hunk 3]
```diff
@@ -403,4 +422,262 @@ impl CudaServerKey {
...
+        }
+
+        if divisor < Scalar::ZERO {
+            self.neg(&quotient, streams)
+        } else {
+            quotient
+        }
+    }
+
+    pub fn signed_scalar_div<Scalar>(
```
Comment (pdroalves) [inline tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs:578]: We should have an async variant for this.

File: tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs [hunk 4]
```diff
@@ -403,4 +422,262 @@ impl CudaServerKey {
...
+                tmp_numerator = numerator.duplicate_async(streams);
+                self.full_propagate_assign_async(&mut tmp_numerator, streams);
+            }
+            &tmp_numerator
+        };
+
+        self.unchecked_signed_scalar_div(numerator, divisor, streams)
+    }
+
+    pub fn unchecked_signed_scalar_div_rem<Scalar>(
```
Comment (pdroalves) [inline tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs:602]: We should have an async variant for this.

File: tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs [hunk 5]
```diff
@@ -403,4 +422,262 @@ impl CudaServerKey {
...
+        let quotient = self.unchecked_signed_scalar_div(numerator, divisor, streams);
+
+        // remainder = numerator - (quotient * divisor)
+        let tmp = self.unchecked_scalar_mul(&quotient, divisor, streams);
+        let remainder = self.sub(numerator, &tmp, streams);
+
+        (quotient, remainder)
+    }
+
+    pub fn signed_scalar_div_rem<Scalar>(
```
Comment (pdroalves) [inline tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs:621]: We should have an async variant for this.

File: tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs [hunk 6]
```diff
@@ -403,4 +422,262 @@ impl CudaServerKey {
...
+                tmp_numerator = numerator.duplicate_async(streams);
+                self.full_propagate_assign_async(&mut tmp_numerator, streams);
+            }
+            &tmp_numerator
+        };
+
+        self.unchecked_signed_scalar_div_rem(numerator, divisor, streams)
+    }
+
+    pub fn unchecked_signed_scalar_rem<Scalar>(
```
Comment (pdroalves) [inline tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs:645]: We should have an async variant for this.

File: tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs [hunk 7]
```diff
@@ -403,4 +422,262 @@ impl CudaServerKey {
...
+    where
+        Scalar: SignedReciprocable + ScalarMultiplier + DecomposableInto<u8> + CastInto<u64>,
+        <<Scalar as SignedReciprocable>::Unsigned as Reciprocable>::DoublePrecision: Send,
+    {
+        let (_, remainder) = self.unchecked_signed_scalar_div_rem(numerator, divisor, streams);
+
+        remainder
+    }
+
+    pub fn signed_scalar_rem<Scalar>(
```
Comment (pdroalves) [inline tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs:660]: We should have an async variant for this.

File: tfhe/src/integer/gpu/server_key/radix/scalar_shift.rs [hunk 8]
```diff
@@ -565,4 +565,76 @@ impl CudaServerKey {
         };
         stream.synchronize();
     }
+
+    /// # Safety
+    ///
+    /// - `stream` __must__ be synchronized to guarantee computation has finished, and inputs must
+    ///   not be dropped until stream is synchronised
+    pub unsafe fn unchecked_scalar_right_shift_logical_assign_async<Scalar, T>(
```
Comment (pdroalves) [inline tfhe/src/integer/gpu/server_key/radix/scalar_shift.rs:573]: Why this method is being added in this PR?

**Review by pdroalves — CHANGES_REQUESTED** [review]
Hey Agnes!

I was not able to review the HL API. Regarding the integer part, I missed the `assign_async`/`async` logic we always use for those operators.

File: tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs (hunk 2 above)
Comment (agnesLeroy) [reply tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs:49]: I made the changes in this PR so everything is consistent.

File: tfhe/src/integer/gpu/server_key/radix/scalar_shift.rs (hunk 8 above)
Comment (agnesLeroy) [reply tfhe/src/integer/gpu/server_key/radix/scalar_shift.rs:573]: It was convenient to add it, it's called in `unchecked_scalar_right_shift_logical_assign_async`.

**Comment by agnesLeroy** [comment]
@pdroalves I modified the PR, I changed also the cast functions to have sync/async variants. The PR is significantly bigger now but I think this improves things a lot, thank you very much for your review.

**Comment by tmontaigu** [comment]
The hlapi part looks good

File: tfhe/src/core_crypto/gpu/entities/lwe_ciphertext_list.rs [hunk 9]
```diff
@@ -56,7 +69,6 @@ impl<T: UnsignedInteger> CudaLweCiphertextList<T> {
         unsafe {
```
Comment (pdroalves) [inline tfhe/src/core_crypto/gpu/entities/lwe_ciphertext_list.rs:69]: We don't need this unsafe clause anymore.

File: tfhe/src/integer/server_key/radix_parallel/scalar_div_mod.rs [hunk 10]
```diff
@@ -135,6 +135,7 @@ pub trait SignedReciprocable:
     type DoublePrecision: DecomposableInto<u8>
         + ScalarMultiplier
         + CastFrom<<Self::Unsigned as Reciprocable>::DoublePrecision>
+        + CastInto<u64>
```
Comment (pdroalves) [inline tfhe/src/integer/server_key/radix_parallel/scalar_div_mod.rs:138]: Do we want to modify CPU's code?

**Review by pdroalves — CHANGES_REQUESTED** [review]
Seems good! Just added two minor comments. There is a forgotten unsafe clause.

File: tfhe/src/integer/server_key/radix_parallel/scalar_div_mod.rs (hunk 10 above)
Comment (agnesLeroy) [reply tfhe/src/integer/server_key/radix_parallel/scalar_div_mod.rs:138]: I have to add it to implement scalar_mul_high actually.

**Review by pdroalves — APPROVED** [review]

**Pull request merged**